enhancements/oc/cli-manager.md

```diff
@@ -0,0 +1,231 @@
+Some current examples are `oc`, `kubectl`, `odo`, `istio`, `tekton`, and `knative`.  It is difficult for users to discover what tools exist,
+where to download them from and which version they should download. We should simplify as much as possible the interaction
+of services on OpenShift. We need a mechanism for providing and consuming tools that is simple to add on to as new plugins are
+developed from a variety of sources - and this should be specific for each cluster and available with disconnected installs.
+
+### Goals
+
+* No new form of binary distribution or binary creation will be proposed, because we have an existing structure at Red Hat.
+RPMs or images are the only options, and images must be deployed by the RH pipeline via operators. This proposal is for delivering
+plugins via images, because this will enable offering plugins offline through an existing image registry.
+* Plugin owners must be able to easily distribute their binaries
+* Allow cluster-admins to control which plugins are offered to users
+  * This proposal is not concerned with _which_ plugins will be managed, as that is decided by cluster-admins
+* Controller for registering CRs with an API for listing, extracting and downloading plugins
+  * An API that generates an index that `krew` can consume via its custom index feature
+  * The index and binaries will served from a single route and service within the cluster
+* The `krew` client package will be vendored into `oc` for usage as `oc krew install <plugin>`
+
+### Non-Goals
+
+* The Controller will not build or package plugins
+* No recommending/limiting which plugins are served by the API
+* `krew` will not create or update the Plugin CRs, those will be managed by cluster-admins
+
+## Proposal
+
+Each component wishing to provide customers with their plugins will build and publish images via a trusted image registry
+and create a Plugin CR to provide an image name and the file path within that image for the binaries.
+Clients (i.e. `krew`) will read the index and download binaries from the Controller. The Controller is responsible for building the index from CRs, for pulling the images,
+extracting the binaries, and serving them to clients.
+
+* `krew` and `krew` plugins are upstream projects that Kubernetes users are already familiar with
+* A `krew`-compatible custom index can provide available plugins for a cluster in disconnected environments
+* The index will be served by a Controller with its contents managed by cluster-admins via CRs
+* The binaries will also be served by the Controller that will pull images from a trusted image registry and extract the binaries
+* With `krew index add https://someother-third-party-index` we won't limit users from adding their own index with whatever plugins they want
+* The controller could be optional, since connected environments can use the default `krew` index out-of-the box
```

Comment:
@deejross the biggest missing element in this proposal is the discussion whether we want that controller to be part of every default OCP installation or rather an optional element. I think it's crucial before we make final decision which direction we want to go. 

Pros for built-in:
- available from day 1 for all clusters
- all binaries including the built-ins can be served using single unified mechanism
Cons for built-in
- additional controller running by default

Pros for opt-in:
- less resource usage by default 
Cons for opt-in:
- yet another operator to manage and mirror, especially in disconnected envs. 
- separate mechanism for built-in and 3rd party binaries
- missing plugin CRs for optional elements installed before this controller is installed. 

I'm personally leaning towards buit-in, and if I don't see any objections I'd like to pursuit that one.

Reply:
The bar for adding an "always present" component is very very high. Close enough to insurmountable that we are considering making existing always-present components optional #922. I don't think this is going to meet that bar. For example, consider all of the disconnected deployments, workload-only clusters, and clusters consumed by kcp. Users won't be interacting with those to get their clients.

Reply:
other discussion from the slack thread(paraphrased):

maciej: this is replacing an existing component (console cli downloader server) so it's not a net-add to the core

bparees: once we have #922 we could make this component optional, so putting it in the core in 4.11 (when we'll have that capability to disable, hopefully) might be an acceptable middle-ground, with a plan to move it to OLM in the future when OLM is capable of the full "cluster-lifecycling/day0" story.

Reply:
@deejross I'm missing the section I've mentioned above describing options and which one was picked any why.

Reply:
@deejross can you replace this sentence and below with something like:
```
* There are two possible approaches for this new controller:
1. Controller should be installed by default...
pros... cons...
2. Controller being optionally installed...
pros... cons...
Given above arguments it is reasonable to include the controller by default.
```